### ISSUE TITLE
key shortcut - bad persistent state?

### ISSUE THREAD

**Issue by TurboGit** [darktable-org/darktable#10898]
**Describe the bug/issue**

Define a shortcut for 'e' + horizontal-mouse move to control exposure. Use it and move up/down, the horizontal move is not working anymore.

**To Reproduce**
_Please provide detailed steps to reproduce the behaviour, for example:_
1. Go 'darkroom'
2. Click on 'e' + mouse-move-horizontal
3. All good this increase the exposure, do not release 'e' key
4. Move mouse up, message on top 'e, left-click, vertical, up' not assigned, OK that's the case, do not release 'e' key
3. Move mouse horizontally again
4. See error, 'e, horizontal, up not assigned'

**Expected behavior**

The Last step should increase the exposure. The issue is that as soon as the mouse is not move correctly you have to release everything and try again.

To me it looks like some state are not reset, in point 6 above the message is having the 'up' info which is suspicious.

**Platform**

* darktable version : master
* OS                : GNU/Linux
* Linux - Distro    : GNU/Debian - sid

**Comment by TurboGit** [comment]
I have set 3.8.1 as target release, but in fact it all depends about the fix (if too intrusive we can move to 4.0).

**Comment by dterrahe** [comment]
I'm afraid I'm not able to reproduce, so let's work through the exact steps together. Could you please paste all entries from your shortcutrc starting with "e"?

> 4\. message on top 'e, left-click, vertical, up' not assigned

Did you left-click with mouse at any point? Otherwise this is already wrong. 

> the message is having the 'up' info which is suspicious

That's not the suspicious part; for any unmapped move, the error would mention the direction. You can map up and down moves to different actions.

> it looks like some state are not reset

Rather, if you make 3 consecutive unmapped moves, the state gets reset. This is to allow dealing with keys getting "stuck", which seems to happen to some people. However, normally what happens is if you keep "e" pressed, gtk will send repeating keystrokes, so you'll soon get back to "e is being pressed" state again. However, if you had a more complex sequence (for example press e and double-click), you may lose the last part of the sequence.

**Comment by TurboGit** [comment]
> Did you left-click with mouse at any point? Otherwise this is already wrong.

Yes, the move I have registered is 'e' + left-click + horizontal-move (I see that my report is indeed missing this, sorry).

Isn't this possible?

I have change the registered shortcut to just : 'e' + horizontal-move and all is working fine.

Maybe I missed something and I must say that I have not read the documentation :(

**Comment by dterrahe** [comment]
Ah, OK, then yes, you are experiencing the drawback of the workaround I implemented to try to help people who reported "stuck keys". Sometimes they seemed to get stuck in a state where the code believed it had received a key press and was sending move shortcuts for each mouse move until that key would be released (which wouldn't happen, because nothing had really been pressed in the first place). I never completely understood this case completely because it was either Mac related or just too random for people to report reproducing steps. The workaround is that if 3 consecutive unmapped shortcuts are received, the  whole state is reset, including your mouse click. As I tried to explain above, since a pressed key will send repeating gtk events, you will quickly get back to the "e" pressed state again, but not to "e+left-click". So now you get simple e+horizontal move shortcuts, which you haven't specified a mapping for...

Hope that explains it?

I'm thinking about implementing a different break-pressed-key-state solution, but would need to make it fit well with any solution to #10817, which is tricky.

So yes, you can map e+click+move, but if you mix it with unmapped moves it gets messy. If you have a mapping for both horizontal and vertical moves, you won't run into this problem.

In that context, have you tried fallbacks? (maybe you don't like them). Then, you just need to map e+click to exposure. You'll automatically get:
1. e+click: popup where you can enter value.
2. e+click+scroll (hard to do): normal slider adjustment
3. e+click+horizontal move: slow slider adjustment
4. e+click+vertical move: fast slider adjustment
The issue you are running into here is less likely to happen then.

**Comment by TurboGit** [comment]
> Hope that explains it?

Yes, thanks!

I didn't tried the fallbacks. I'll see if I can get something out of this. Should I close or let this open pending #10817?
